### PR TITLE
Log in, log out

### PR DESCRIPTION
**Pull request by devtuananh**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: user&.user.authenticate params[:session][:password]

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  def logged_in?
+    current_user.present?
+  end
+
+  def current_user
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by id: user_id
+    elsif user_id = cookies.signed[:user_id]
+      user = User.find_by id: user_id
+      if user && user.authenticated?(:remember, cookies[:remember_token])
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: user&.user.authenticated? :remember, cookies[:remember_token]

File: app/models/user.rb [hunk 3]
```diff
@@ -1,14 +1,54 @@
...
-  validates :name, presence: true, length: {maximum: Settings.validates.name_maximum}
+  has_secure_password
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  validates :email, presence: true, uniqueness: true,
-    length: {maximum: Settings.validates.email_maximum}, format: {with: VALID_EMAIL_REGEX}
-  validates :password, presence: true, length:
-    {minimum: Settings.validates.password_minimum}, allow_nil: true
+  validates :email, format: {with: VALID_EMAIL_REGEX},
+    length: {maximum: Settings.validates.email_maximum},
+    presence: true, uniqueness: true,
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove last ,

File: app/models/user.rb [hunk 4]
```diff
@@ -1,14 +1,54 @@
...
+
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove this empty line

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+module SessionsHelper
+
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: Please remove this this empty line

Comment (devtuananh) [reply app/helpers/sessions_helper.rb]: yes

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.user.authenticate params[:session][:password]
+      log_in user
+      params[:session][:remember_me] == Settings.check_session ?
+        remember(user) : forget(user)
+      redirect_to root_url
```
Comment (phucynwa) [inline app/controllers/sessions_controller.rb]: we should you root_path

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    else
+      flash.now[:danger] = t ".invalid"
+      render :new
+    end
+  end
+
+  def destroy
+    log_out
+    redirect_to root_url
```
Comment (phucynwa) [inline app/controllers/sessions_controller.rb]: same

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    session[:user_id] = user.id
+  end
+
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def current_user? user
```
Comment (hungnh103) [inline app/helpers/sessions_helper.rb]: em chuyển method này vào model User nhé

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.user.authenticate params[:session][:password]
+      log_in user
+      params[:session][:remember_me] == Settings.check_session ?
```
Comment (phucynwa) [inline app/controllers/sessions_controller.rb]: use `remember_me_checked` instead `check_session`

File: app/helpers/sessions_helper.rb (hunk 8 above)
Comment (devtuananh) [reply app/helpers/sessions_helper.rb]: ok anh

**Pull request merged**